File: src/System/Mesos/Raw/ExecutorInfo.hs
module System.Mesos.Raw.ExecutorInfo where
import           System.Mesos.Internal
import           System.Mesos.Raw.CommandInfo
import           System.Mesos.Raw.ContainerInfo
import           System.Mesos.Raw.DiscoveryInfo
import           System.Mesos.Raw.ExecutorId
import           System.Mesos.Raw.FrameworkId
import           System.Mesos.Raw.Resource

type ExecutorInfoPtr = Ptr ExecutorInfo

foreign import ccall unsafe "ext/types.h toExecutorInfo" c_toExecutorInfo
  :: ExecutorIDPtr
  -> FrameworkIDPtr
  -> CommandInfoPtr
  -> ContainerInfoPtr
  -> Ptr ResourcePtr  -- ^ resources
  -> CInt      -- ^ resource count
  -> Ptr CChar -- ^ name
  -> CInt      -- ^ name length
  -> Ptr CChar -- ^ source
  -> CInt      -- ^ source length
  -> DiscoveryInfoPtr -- ^ discoveryInfo
  -> IO ExecutorInfoPtr

foreign import ccall unsafe "ext/types.h fromExecutorInfo" c_fromExecutorInfo
  :: ExecutorInfoPtr
  -> Ptr ExecutorIDPtr
  -> Ptr FrameworkIDPtr
  -> Ptr CommandInfoPtr
  -> Ptr ContainerInfoPtr
  -> Ptr (Ptr ResourcePtr) -- ^ resources
  -> Ptr CInt -- ^ resources count
  -> Ptr (Ptr CChar) -- ^ name
  -> Ptr CInt -- ^ name length
  -> Ptr (Ptr CChar) -- ^ source
  -> Ptr CInt -- ^ source length
  -> Ptr (Ptr DiscoveryInfo) -- ^ discovery info
  -> IO ()

foreign import ccall unsafe "ext/types.h destroyExecutorInfo" c_destroyExecutorInfo
  :: ExecutorInfoPtr
  -> IO ()

instance CPPValue ExecutorInfo where

  marshal i = do
    eidP <- cppValue $ executorInfoExecutorId i
    fidP <- cppValue $ executorInfoFrameworkId i
    ciP <- cppValue $ executorInfoCommandInfo i
    ctrP <- case executorInfoContainerInfo i of
              Nothing -> return nullPtr
              Just ctr -> cppValue ctr
    rps <- mapM cppValue $ executorInfoResources i
    (np, nl) <- maybeCString $ executorInfoName i
    (sp, sl) <- maybeCString $ executorInfoSource i
    (rs, rLen) <- arrayLen rps
    discp <- maybe (return nullPtr) cppValue $ executorInfoDiscover i
    liftIO $ c_toExecutorInfo eidP fidP ciP ctrP rs (fromIntegral rLen) np (fromIntegral nl) sp (fromIntegral sl) discp

  unmarshal ip = do
    eidP <- alloc
    fidP <- alloc
    ciP <- alloc
    cntP <- alloc
    rpP <- alloc
    rpL <- alloc
    enP <- alloc
    enL <- alloc
    esP <- alloc
    esL <- alloc
    discP <- alloc
    poke enP nullPtr
    poke esP nullPtr
    poke cntP nullPtr
    poke rpP nullPtr
    poke discP nullPtr
    liftIO $ c_fromExecutorInfo ip eidP fidP ciP cntP rpP rpL enP enL esP esL discP
    eid <- unmarshal =<< peek eidP
    fid <- unmarshal =<< peek fidP
    ci <- unmarshal =<< peek ciP
    cnt <- peekMaybeCPP cntP
    rl <- peek rpL
    rp <- peek rpP
    rps <- peekArray' (rp, fromIntegral rl)
    rs <- mapM unmarshal rps
    en <- peekMaybeBS enP enL
    es <- peekMaybeBS esP esL
    disc <- peekMaybeCPP discP
    return $ ExecutorInfo eid fid ci cnt rs en es disc

  destroy = c_destroyExecutorInfo

  equalExceptDefaults (ExecutorInfo id fid ci cnt rs n s d) (ExecutorInfo id' fid' ci' cnt' rs' n' s' d') =
    id == id' && fid == fid' && equalExceptDefaults ci ci' &&
    and (zipWith equalExceptDefaults rs rs') && n == n'&&
    s == s' && d == d' && cnt == cnt'

File: ext/types/executor_info.cpp
#include <iostream>
#include "types.h"

using namespace mesos;

ExecutorInfoPtr toExecutorInfo(ExecutorIDPtr executorID,
			       FrameworkIDPtr frameworkID,
			       CommandInfoPtr commandInfo,
			       ContainerInfoPtr containerInfo,
			       ResourcePtr* resources,
			       int resourceLen,
			       char* name,
			       int nameLen,
			       char* source,
			       int sourceLen,
             DiscoveryInfoPtr discovery)
{
  ExecutorInfoPtr info = new ExecutorInfo();
  *info->mutable_executor_id() = *executorID;
  *info->mutable_framework_id() = *frameworkID;
  *info->mutable_command() = *commandInfo;
  if (containerInfo != NULL)
    *info->mutable_container() = *containerInfo;

  for (int i = 0; i < resourceLen; ++i)
    *info->add_resources() = *resources[i];

  if (name != NULL)
    info->set_name(name, nameLen);

  if (source != NULL)
    info->set_source(source, sourceLen);

  if (discovery != NULL)
    *info->mutable_discovery() = *discovery;

  return info;
}

void fromExecutorInfo(ExecutorInfoPtr info,
		      ExecutorIDPtr* executorID,
		      FrameworkIDPtr* frameworkID,
		      CommandInfoPtr* commandInfo,
		      ContainerInfoPtr* containerInfo,
		      ResourcePtr** resources,
		      int* resourcesLen,
		      char** name,
		      int* nameLen,
		      char** source,
		      int* sourceLen,
          DiscoveryInfoPtr* discovery)
{
  *executorID = info->mutable_executor_id();

  *frameworkID = info->mutable_framework_id();

  *commandInfo = info->mutable_command();

  *resources = info->mutable_resources()->mutable_data();
  *resourcesLen = info->resources_size();

  if (info->has_container())
    *containerInfo = info->mutable_container();

  if (info->has_name())
    {
      std::string* n = info->mutable_name();
      *name = (char*) n->data();
      *nameLen = n->size();
    }

  if (info->has_source())
    {
      std::string* s = info->mutable_source();
      *source = (char*) s->data();
      *sourceLen = s->size();
    }

  if (info->has_discovery())
    {
      *discovery = info->mutable_discovery();
    }
}

void destroyExecutorInfo(ExecutorInfoPtr info)
{
  delete info;
}
